File: app/src/main/java/bivachi/agustin/calculator/controller/CalculatorController.java
package bivachi.agustin.calculator.controller;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Locale;

import bivachi.agustin.calculator.model.PendingBinaryOperation;

/**
 * Created by agustinbivachi on 26/4/17.
 */
public class CalculatorController {

  private Double accumulator;
  private PendingBinaryOperation pendingOperation;

  public CalculatorController() {
    this.accumulator = 0.0;
    this.pendingOperation = new PendingBinaryOperation();
  }

  public void setAccumulator(String display) {
    this.accumulator = parseStringToNumber(display);
  }

  public String getResult(){
    if (accumulator!=null){
      return formatNumberToString(accumulator);
    }else {
      reset();
      return "Error";
    }
  }

  public void performOperation(String currentSymbol) {
      executePendingBinaryOperation();
      pendingOperation.setOperationValues(accumulator,currentSymbol);
  }

  private void executePendingBinaryOperation() {
    if (pendingOperation.getCurrentOperationToPerform() != null) {
      accumulator = pendingOperation.getCurrentOperationToPerform().calculate(pendingOperation.getFirstOperand(), accumulator);
      pendingOperation.resetPendingBinaryOperation();
    }
  }

  public void reset() {
    this.accumulator = 0.0;
    pendingOperation.resetPendingBinaryOperation();
  }

  public String formatNumberToString(Double result) {
    NumberFormat nf = NumberFormat.getNumberInstance(Locale.GERMAN);
    DecimalFormat decimalFormat = (DecimalFormat)nf;
    return decimalFormat.format(result);
  }

  private Double parseStringToNumber(String display) {
    return Double.parseDouble(display.replace(".","").replace(",",".").replace("Error","0"));
    //replace methods are to handle parse errors
  }

}

File: app/src/main/java/bivachi/agustin/kotlintest/controller/CalculatorController.kt
package bivachi.agustin.kotlintest.controller

import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.Locale

import bivachi.agustin.kotlintest.model.PendingBinaryOperation

/**
 * Created by agustinbivachi on 26/4/17.
 */
class CalculatorController {

    private var accumulator: Double?
    private val pendingOperation: PendingBinaryOperation

    init {
        this.accumulator = 0.0
        this.pendingOperation = PendingBinaryOperation()
    }

    fun setAccumulator(display: String) {
        this.accumulator = parseStringToNumber(display)
    }

    val result: String
        get() {
            if (accumulator != null || accumulator!!.toString() != "Infinity") {
                return formatNumberToString(accumulator)
            } else {
                reset()
                return "Error"
            }
        }

    fun performOperation(currentSymbol: String) {
        executePendingBinaryOperation()
        pendingOperation.setOperationValues(accumulator, currentSymbol)
    }

    private fun executePendingBinaryOperation() {
        if (pendingOperation.currentOperationToPerform != null) {
            accumulator = pendingOperation.currentOperationToPerform!!.operation(pendingOperation.firstOperand!!, accumulator!!)
            pendingOperation.resetPendingBinaryOperation()
        }
    }

    fun reset() {
        this.accumulator = 0.0
        pendingOperation.resetPendingBinaryOperation()
    }

    fun formatNumberToString(result: Double?): String {
        val nf = NumberFormat.getNumberInstance(Locale.GERMAN)
        val decimalFormat = nf as DecimalFormat
        return decimalFormat.format(result)
    }

    private fun parseStringToNumber(display: String): Double {

        return java.lang.Double.parseDouble(display.replace(".", "").replace(",", ".").replace("Error", "0"))
        //replace methods are to handle parse errors
    }


}
